### PR TITLE
Use micro RPC for Wasm API

### PR DESCRIPTION
**Pull request by tiziano88**
Move away from the concept of extensions (which were essentially individual methods addressed by a number), in favour of using a single micro RPC service to define all the methods at once in a type safe way.

This requires updating the SDK in each language, and it is a breaking change for existing Wasm module, so it will require some coordination before merging internally, especially for the C++ SDK (which will have to be implemented manually as we don't have C++ bindings for micro RPC yet).

The API object probably has more locking than necessary, but it should be easier to remove later as part of a follow up commit.

Ref #3900

### CODE REVIEW

**Comment by tiziano88** [comment]
FYI @mariaschett @pmcgrath17 (no review required, but if you have comments feel free to send them over)

**Comment by pmcgrath17** [comment]
So with this change (and subsequent change), we should be able to reference the write_response, read_request, and invoke oak functions by using the micro RPC interface?

**Comment by tiziano88** [comment]
> So with this change (and subsequent changes), we should be able to reference the write_response, read_request, and invoke oak functions by using the micro RPC interface?

I think you are thinking about the micro RPC interface from the untrusted launcher? if so, no, this is not it. This makes the interface between the Wasm module and the Oak Functions runtime also use micro RPC, instead of the exported function that we used to have. This way, we can add more functionality to the Wasm API over time without breaking compatibility at the Wasm level. For instance we could have prefix search for lookups, or more complex ways of querying the lookup data, etc.

File: oak_functions/wasm/src/lib.rs [hunk 1]
```diff
@@ -563,15 +479,10 @@ where
             &format!("running Wasm module completed with result: {:?}", result),
         );
 
-        // Terminate the extensions.
-        store
-            .data_mut()
-            .extensions
-            .values_mut()
-            .try_for_each(|e| e.terminate())?;
+        let response_bytes = response.lock().clone();
```
Comment (conradgrobler) [inline oak_functions/wasm/src/lib.rs:482]: It would be nice as a future improvement to not have to clone the bytes and remove the spinlock.

**Review by conradgrobler — APPROVED** [review]
Nice.

**Pull request merged**